r/R/parquet.R

```diff
@@ -15,6 +15,272 @@
+      private$.set(table, write_statistics, "write_statistics", is.logical,
+        parquet___ArrowWriterProperties___Builder__default_write_statistics,
+        parquet___ArrowWriterProperties___Builder__set_write_statistics
+      )
+    },
+
+    set_data_page_size = function(data_page_size) {
+      parquet___ArrowWriterProperties___Builder__data_page_size(self, data_page_size)
+    }
+  ),
+
+  private = list(
+    .set = function(table, value, name, is, default, multiple) {
+      msg <- paste0("unsupported ", name, "= specification")
+      assert_that(is(value), msg = msg)
+      column_names <- names(table)
+      if (is.null(given_names <- names(value))) {
+        if (length(value) == 1L) {
+          default(self, value)
+        } else if (length(value) == length(column_names)) {
+          multiple(self, column_names, value)
+        }
+      } else if(all(given_names %in% column_names)) {
+        multiple(self, given_names, value)
+      } else {
+        abort(msg)
+      }
+    }
+  )
+
+)
+
+ParquetWriterProperties$create <- function(table, version = NULL, compression = NULL, compression_level = NULL, use_dictionary = NULL, write_statistics = NULL, data_page_size = NULL) {
+  if (is.null(version) && is.null(compression) && is.null(compression_level) && is.null(use_dictionary) && is.null(write_statistics) && is.null(data_page_size)) {
+    shared_ptr(ParquetWriterProperties, parquet___default_writer_properties())
+  } else {
+    builder <- shared_ptr(ParquetWriterPropertiesBuilder, parquet___WriterProperties___Builder__create())
+    if (!is.null(version)) {
+      builder$set_version(version)
+    }
+    if (!is.null(compression)) {
+      builder$set_compression(table, compression = compression)
+    }
+    if (!is.null(compression_level)) {
+      builder$set_compression_level(table, compression_level = compression_level)
+    }
+    if (!is.null(use_dictionary)) {
+      builder$set_dictionary(table, use_dictionary)
+    }
+    if (!is.null(write_statistics)) {
+      builder$set_write_statistics(table, write_statistics)
+    }
+    if (!is.null(data_page_size)) {
+      builder$set_data_page_size(data_page_size)
+    }
+    shared_ptr(ParquetWriterProperties, parquet___WriterProperties___Builder__build(builder))
+  }
+}
+
+ParquetFileWriter <- R6Class("ParquetFileWriter", inherit = Object,
+  public = list(
+    WriteTable = function(table, chunk_size) {
+      parquet___arrow___FileWriter__WriteTable(self, table, chunk_size)
+    },
+    Close = function() {
+      parquet___arrow___FileWriter__Close(self)
+    }
+  )
+
+)
+ParquetFileWriter$create <- function(
+  schema,
+  sink,
+  properties = ParquetWriterProperties$create(),
+  arrow_properties = ParquetArrowWriterProperties$create()
+) {
+  unique_ptr(
+    ParquetFileWriter,
+    parquet___arrow___ParquetFileWriter__Open(schema, sink, properties, arrow_properties)
+  )
+}
+
+#' Write Parquet file to disk
```

Comment:
IMO `read_parquet()` and `write_parquet()` should be at the top of this file since they're the main interface we expose, and the R6 classes should follow.